// @node-red/nodes/core/agora/10-MediaStream.js
const axios = require("axios");
const mqtt = require("mqtt");

module.exports = function(RED) {
    function MediaStreamNode(config) {
        RED.nodes.createNode(this,config);
        var node = this;
        node.on("input", function(msg, send, done){
            let payload;
            if (!msg.payload){
                payload = {
                    type: "upstream",
                };
            }else{
                payload = JSON.parse(msg.payload);
            }
            payload.streamOptions = JSON.parse(JSON.stringify(config.streamOptions));
            RED.util.setMessageProperty(msg, "topic", "/agora/upstream/" + config.deviceName, true);
            RED.util.setMessageProperty(msg, "payload", payload, true);
            // msg.payload = JSON.stringify(payload)
            send(msg);
            // console.error("Sent!!!", msg);
            done()
        });
        setTimeout(()=>{
            const msg = {};
            node.emit("input", msg);
            setTimeout(()=>{
                const gMsg = getGlobalMsg(msg);
                sendToEaseMob(gMsg);
                // console.error("global.msgs", JSON.stringify(global.msgs, null, 2));
            }, 50);
        }, 50)
    }
    RED.nodes.registerType("media-stream",MediaStreamNode);
}

global.mqttClient = null
global.getMqttClient = async function(){
    if (mqttClient){
        return mqttClient;
    }else{
        const res = await axios({
            method: 'POST',
            url: `http://a1.easemob.com/1128210519085787/fangfa2/token`,
            data: JSON.stringify({
                grant_type: "password",
                username: "mqttuser2",
                password:"mqttuser2"
            }, null, 2),
            headers: {
                'Content-Type': 'application/json'
            }
        });
        if (mqttClient){
            return mqttClient
        }else{
            global.mqttClient  = mqtt.connect('mqtts://0ljii0.cn1.mqtt.chat:1884', {
                protocolVersion: 5,
                clientId: "mqttuser2@0ljii0",
                username: "mqttuser2",
                password: res.data.access_token
            })
            mqttClient.on('connect', function () {
                console.log("sendToEaseMob Connected");
            });
            mqttClient.on('error', function(err){
                console.error(err)
            })
            return mqttClient;
        }
    }
}

async function sendToEaseMob(msg){
    const client = await getMqttClient();
    console.error("Send message", msg.topic, JSON.stringify(msg.payload, null, 2));
    msg.payload.requestId = Date.now();
    client.publish(msg.topic, JSON.stringify(msg.payload, null, 2), {
        retain: true
    });
    client.publish(msg.topic.replace("upstream", "downstream"), null, {
        retain: true
    });
    console.log("Publish to topic", msg.topic, "\t\t\t", msg.payload);

}
global.msgs = {}
let cnt = 100000;
global.getGlobalMsg = function(msg){
    if (!msg._msgid){
        msg._msgid = cnt++
    }
    if (!global.msgs[msg._msgid]){
        global.msgs[msg._msgid] = msg;
    }
    return global.msgs[msg._msgid];
}
